src/lib/openjp2/vector.c

```diff
@@ -0,0 +1,90 @@
+/*
+ * The copyright in this software is being made available under the 2-clauses 
+ * BSD License, included below. This software may be subject to other third 
+ * party and contributor rights, including patent rights, and no such rights
+ * are granted under this license.
+ *
+ * Copyright (c) 2015, Aaron Boxer
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS `AS IS'
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "opj_includes.h"
+
+const OPJ_INT32 VECTOR_INITIAL_CAPACITY = 64;
+
+static void opj_vec_double_capacity_if_full(opj_vec_t *vec);
+
+void opj_vec_init(opj_vec_t *vec) {
+	vec->size = 0;
+	vec->capacity = VECTOR_INITIAL_CAPACITY;
+	vec->data = opj_malloc(sizeof(void*) * vec->capacity);
+}
+
+void opj_vec_append(opj_vec_t *vec, void* value) {
+	opj_vec_double_capacity_if_full(vec);
+	vec->data[vec->size++] = value;
+}
+
+void* opj_vec_get(opj_vec_t *vec, OPJ_INT32 index) {
+	if (index >= vec->size || index < 0) {
+		printf("Index %d out of bounds for vec of size %d\n", index, vec->size);
+		return NULL;
+	}
+	return vec->data[index];
+}
+
+void* opj_vec_back(opj_vec_t *vec) {
+	if (!vec || !vec->data || !vec->size)
+		return NULL;
+	return opj_vec_get(vec, vec->size - 1);
+}
+
+void opj_vec_set(opj_vec_t *vec, OPJ_INT32 index, void* value) {
+	while (index >= vec->size) {
+		opj_vec_append(vec, NULL);
+	}
+	vec->data[index] = value;
+}
+
+static void opj_vec_double_capacity_if_full(opj_vec_t *vec) {
+	if (vec->size >= vec->capacity) {
+		vec->capacity *= 2;
+		vec->data = opj_realloc(vec->data, sizeof(void*) * vec->capacity);
```

Comment:
`opj_realloc` is not checked. No error returned if `NULL` (& current `vec->data` overriden to `NULL` but not freed in this case).